_studio/mfx_lib/decode/vp9/src/mfx_vp9_dec_decode_hw.cpp

```diff
@@ -85,6 +87,143 @@ bool CheckHardwareSupport(VideoCORE *p_core, mfxVideoParam *p_video_param)
+
+    void UnLockResources(const UMC_VP9_DECODER::VP9DecoderFrame & frame) const
+    {
+        // unlock curent frame for decode
+        VP9_CHECK_AND_THROW((m_frameAllocator->DecreaseReference(frame.currFrame) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+        // unlock frame for copy
+        if (frame.show_existing_frame)
+        {
+            VP9_CHECK_AND_THROW((m_frameAllocator->DecreaseReference(frame.ref_frame_map[frame.frame_to_show]) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+            return;
+        }
+        // unlock all references
+        for (mfxI32 ref_index = 0; ref_index < NUM_REF_FRAMES; ++ref_index)
+        {
+            const UMC::FrameMemID mid = frame.ref_frame_map[ref_index];
+            if (mid >= 0)
+            {
+                VP9_CHECK_AND_THROW((m_frameAllocator->DecreaseReference(mid) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+            }
+        }
+    }
+
+    void UpdateRefFrames(UMC_VP9_DECODER::VP9DecoderFrame & frame) const
+    {
+        if (frame.show_existing_frame)
+            return;
+
+        for (mfxI32 ref_index = 0; ref_index < NUM_REF_FRAMES; ++ref_index)
+        {
+            mfxU8 mask = frame.refreshFrameFlags & (1 << ref_index);
+            if (mask != 0 // we should update the reference according to the bitstream data
+
+                // The next condition is for decoder robustness.
+                // After the first keyframe is decoded this frame occupies all ref slots
+                // its mask is 011111111 and the condition "minus one" never hits.
+                // But if the first frame is Inter-frame encoded in the intra-only mode
+                // then we can receive "minus one" ref_frame_map item.
+                // The decoder never should touch these references.
+                // But if the stream is broken then decode may touch not initialized reference
+                // So to prevent crash let's put something as the reference frame.
+                || (frame.ref_frame_map[ref_index] == (UMC::FrameMemID)-1))
+            {
+                VP9_CHECK_AND_THROW((m_frameAllocator->IncreaseReference(frame.currFrame) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+
+                if (frame.ref_frame_map[ref_index] >= 0)
+                    VP9_CHECK_AND_THROW((m_frameAllocator->DecreaseReference(frame.ref_frame_map[ref_index]) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+                frame.ref_frame_map[ref_index] = frame.currFrame;
+
+                frame.sizesOfRefFrame[ref_index].width = frame.width;
+                frame.sizesOfRefFrame[ref_index].height = frame.height;
+            }
+        }
+    }
+
+public:
+    FrameStorage(mfx_UMC_FrameAllocator *frameAllocator):
+        m_framesStorage(),
+        m_frameAllocator(frameAllocator)
+    {
+    }
+
+    void Add(UMC_VP9_DECODER::VP9DecoderFrame & frame)
+    {
+        VP9_CHECK_AND_THROW((frame.currFrame >= 0), MFX_ERR_UNKNOWN);
+
+        UpdateRefFrames(frame);
+        LockResources(frame);
+        CompleteFrames();
+
```

Comment:
Wrong function calls order. First LockResources for the current frame. After the resources are locked we can call UpdateRefFrames which will update the reference frame list for the next frame and this function should be called from the decoder, not from the FrameStorage (this change should break async decoding, did you test it). I believe CompleteFrames should not be called from here. I see the only place where it should be called is the first statement in DecodeFrameCheck, because of guard(mutex) no one will be able to modify FrameStorage.

Reply:
Sure I  tested change on conformance suites